src/routing/bootstrap.rs

```diff
@@ -0,0 +1,653 @@
+            None => self.node.name(),
+        };
+
+        let message = Message::single_src(
+            &self.node.keypair,
+            self.node.age,
+            DstLocation::Direct,
+            Variant::BootstrapRequest(destination),
+            None,
+            None,
+        )?;
+
+        debug!("{} Sending BootstrapRequest to {:?}", self.node, recipients);
+
+        let _ = self.send_tx.send((message.to_bytes(), recipients)).await;
+
+        Ok(())
+    }
+
+    async fn receive_bootstrap_response(&mut self) -> Result<(BootstrapResponse, SocketAddr)> {
+        while let Some((message, sender)) = self.recv_rx.recv().await {
+            match message.variant() {
+                Variant::BootstrapResponse(response) => {
+                    if !self.verify_message(&message, None) {
+                        continue;
+                    }
+
+                    return Ok((response.clone(), sender));
+                }
+                _ => {
+                    trace!(
+                        "{} Useless message {:?} from {}",
+                        self.node,
+                        message,
+                        sender,
+                    );
+                }
+            }
+        }
+
+        error!("{} Message sender unexpectedly closed", self.node);
+        Err(Error::InvalidState)
+    }
+
+    // Change our name to fit the destination section and apply the new age.
+    fn process_relocation(
+        &mut self,
+        elders_info: &EldersInfo,
+        relocate_details: SignedRelocateDetails,
+    ) -> Result<RelocatePayload> {
+        // We are relocating so we need to change our name.
+        // Use a name that will match the destination even after multiple splits
+        let extra_split_count = 3;
+        let name_prefix = Prefix::new(
+            elders_info.prefix.bit_count() + extra_split_count,
+            *relocate_details.destination(),
+        );
+
+        let mut rng = crate::rng::new();
+        let new_keypair = crypto::keypair_within_range(&mut rng, &name_prefix.range_inclusive());
+        let new_name = crypto::name(&new_keypair.public);
+        let age = relocate_details.relocate_details().age;
+        let relocate_payload =
+            RelocatePayload::new(relocate_details, &new_name, &self.node.keypair)?;
+
+        info!("{} Changing name to {}.", self.node, new_name);
+        self.node = Node::with_age(new_keypair, self.node.addr, age);
+
+        Ok(relocate_payload)
+    }
+
+    // Send `JoinRequest` and wait for the response. If the response is `Rejoin`, repeat with the
+    // new info. If it is `Approval`, returns the initial `Section` value to use by this node,
+    // completing the bootstrap.
+    async fn join(
+        mut self,
+        mut elders_info: EldersInfo,
+        mut section_key: bls::PublicKey,
+        relocate_payload: Option<RelocatePayload>,
+    ) -> Result<(Node, Section)> {
+        loop {
```

Comment:
bit concerned as the only break out of the loop is to be `Approved`, otherwise will keeps sending out join request or waiting for a response forever.

Reply:
Same